crates/fluvio-benchmark/src/stats_collector.rs

```diff
@@ -1,162 +1,207 @@
+                        record_bytes: bytes_delta,
+                        records_per_sec,
+                        bytes_per_sec,
+                        latency_avg,
+                        latency_max,
+                    })
+                    .await
+                    .expect("send stats");
+            }
+        });
+        Ok(())
+    }
+
     fn track_latency(
-        num_records: u64,
-        latency_sender: Sender<Histogram<u64>>,
-        rx: async_channel::Receiver<(ProduceOutput, Instant)>,
+        end_sender: Sender<EndProducerStat>,
         histogram: Arc<Mutex<Histogram<u64>>>,
+        event_receiver: Receiver<ProduceCompletionEvent>,
+        stats: Arc<AtomicStats>,
+        start_time: Arc<Mutex<Option<Instant>>>,
     ) {
         spawn(async move {
-            while let Ok((send_out, time)) = rx.recv().await {
-                let hist = histogram.clone();
-                let latency_sender = latency_sender.clone();
-                //spawn(async move {
-                match send_out.wait().await {
-                    Ok(_) => {
-                        let duration = time.elapsed();
+            let hist = histogram.clone();
+            while let Ok(event) = event_receiver.recv().await {
+                match event.metadata.base_offset().await {
+                    Ok(_base_offset) => {
+                        let elapsed = event.created_at.elapsed();
                         let mut hist = hist.lock().await;
-                        hist.record(duration.as_nanos() as u64).expect("record");
-
-                        if hist.len() >= num_records {
-                            latency_sender.send(hist.clone()).await.expect("send");
-                        }
+                        hist.record(elapsed.as_nanos() as u64).expect("record");
                     }
                     Err(err) => {
-                        println!("error sending record: {}", err);
-                        return;
+                        println!("received err: {:?}", err);
                     }
                 }
-                //});
             }
-        });
-    }
 
-    pub(crate) fn calcuate(&mut self) -> Stat {
-        let elapse = self.start_time.elapsed().as_millis();
-        let records_per_sec = ((self.record_send as f64 / elapse as f64) * 1000.0).round();
-        let bytes_per_sec = (self.record_bytes as f64 / elapse as f64) * 1000.0;
-
-        let hist = self.histogram.lock_blocking();
-        let latency_avg = hist.mean() as u64;
-        let latency_max = hist.value_at_quantile(1.0);
-
-        Stat {
-            records_per_sec,
-            bytes_per_sec,
-            _total_bytes_send: self.record_bytes,
-            total_records_send: self.record_send,
-            latency_avg,
-            latency_max,
-            _end: false,
-        }
+            // send end
+            let hist = hist.lock().await;
+            let start_time = start_time.lock().await.expect("start time");
```

Comment:
start time should be tracked by callback?

Reply:
This is the first start time, I can rename it

Reply:
using first start time form callback